base/math.jl

```diff
@@ -841,6 +841,17 @@ function _hypot(x::NTuple{N,<:Number}) where {N}
     end
 end
 
+function _hypot(x::NTuple{N,T}) where {N,T<:IEEEFloat}
+    infT = convert(T, Inf)
+    x = abs.(x) # doesn't change result but enables computational shortcuts
+    any(==(infT), x) && return infT # return Inf even if an argument is NaN
```

Comment:
just use `isinf` here. With https://github.com/JuliaLang/julia/pull/48109 it should be as fast.

Reply:
I'd tried this but the performance difference isn't as negligible as I'd like on longer tuples:
```julia-repl
# use `isinf`
  7.708 ns (0 allocations: 0 bytes)
  7.908 ns (0 allocations: 0 bytes)
  11.111 ns (0 allocations: 0 bytes)
  19.860 ns (0 allocations: 0 bytes)
```
Looking at the `code_native`, the `isinf` version turns into a branch chain (with short-circuit evaluation) while `==(Inf)` (which is valid only because we already `abs`'d the values) results in branchless SIMD.

So there are three options:
1) use `==(Inf)` here
2) use `isinf` here but do not change it (performance penalty)
3) use `isinf` here after changing it to use `==` instead of `===` (I verified this gives the same assembly as the current version)

Since that `isinf` change is less than a day old and the performance difference between `==` and `===` was very small, I'd vote for option 3.

Reply:
== vs === shouldn't be making a difference in whether it simds. that imo is an LLVM issue

Reply:
The issue is that LLVM fails to recognize that `x===c` and `x==c` are equivalent for any value `c` except `-0.0`, `+0.0`, or `NaN`. Thus, it always does `x===c` using integer instructions and `x==c` using float instructions. The relevant x64-64 assembly to the `any(isinf,x)` check:
```
#### with ===
        vmovq   %xmm1, %rcx
        movabsq $.LCPI0_1, %rdx                                                            # +Inf
        vmovsd  (%rdx), %xmm0                   # xmm0 = mem[0],zero       # place +Inf as the possible return value (not relevant to this code section)
        cmpq    %rax, %rcx
        je      .LBB0_6
# %bb.1:                                # %L18.lr.ph
        vpextrq $1, %xmm1, %rdx
        cmpq    %rax, %rdx
        je      .LBB0_6
# %bb.2:                                # %L18.1
        vextracti128    $1, %ymm1, %xmm2
        vmovq   %xmm2, %rsi
        cmpq    %rax, %rsi
        je      .LBB0_6
# %bb.3:                                # %L18.2
        vpextrq $1, %xmm2, %rdi
        cmpq    %rax, %rdi
        je      .LBB0_6

#### with ==
        movabsq $.LCPI0_1, %rax                                                         # +Inf
        vbroadcastsd    (%rax), %ymm0
        vcmpeqpd        %ymm0, %ymm1, %ymm0
        vmovmskpd       %ymm0, %ecx
        testl   %ecx, %ecx
        je      .LBB0_3
```

Reply:
I wish LLVM were smarter here, but we can't have nice things I guess.

Reply:
The branching with `===` disappears if I use `mapreduce(isinf,|,x)` instead of `any(isinf,x)`. However, it still doesn't SIMD and the performance difference doesn't change noticeably (`Inf` never appears in my benchmark evaluation, so the branches should be predicted very well). I don't know why LLVM is letting us down here, except that maybe the integer SIMD instructions don't exist or it isn't as smart about using them.

Any thoughts on the 3 options I listed above? In #48109, we definitely saw that the `===` version of `isinf` was very slightly faster than `==`. To make sure our original benchmarks of `isinf` weren't overly trivial, I benchmarked where `any(isinf,...)` was used for branch selection and it also slightly favored `===`. But here we do see a practical case where `==` results in a notable performance benefit. Honestly, all my intuition would have predicted (incorrectly) that `===` would be at least as good as `==` here since the values are used as integers in the `maximum` that immediately follows.

I can't predict how this would play out across a larger set of practical usages, except that the few other toy benchmarks I've tried tie or favor `===`. My only argument for `==` (beyond the specific case arising in this PR) is that it may be marginally easier for LLVM to optimize if it can someday figure out that it can alternatively use `===` for many comparisons (figuring out the reverse direction seems tougher, or at least not easier).

The evidenced option (and the coward's option) is to keep `isinf` using `===` and this function using `==` even if the discrepancy seems silly.